src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -424,11 +431,39 @@ ClassFieldMap::~ClassFieldMap() {
   delete _fields;
 }
 
+int ClassFieldMap::interfaces_field_count(InstanceKlass* ik) {
+  const Array<InstanceKlass*>* interfaces = ik->transitive_interfaces();
+  int count = 0;
+  for (int i = 0; i < interfaces->length(); i++) {
+    FilteredJavaFieldStream fld(interfaces->at(i));
+    count += fld.field_count();
+  }
+  return count;
+}
+
 void ClassFieldMap::add(int index, char type, int offset) {
   ClassFieldDescriptor* field = new ClassFieldDescriptor(index, type, offset);
   _fields->append(field);
 }
 
+int ClassFieldMap::add_instance_fields(InstanceKlass* ik, int start_index) {
+  // add super fields first
+  InstanceKlass* super_klass = ik->java_super();
+  if (super_klass != nullptr) {
+    start_index += add_instance_fields(super_klass, start_index);
```

Comment:
Does hotspot have any rules against potentially very deep recursion that can overflow the stack?

Reply:
I looked at hotspot code and don't see "implementation limits" for level of inheritance, so it looks like a valid concern.
Need to reimplement this without recursion.

Reply:
Fixed.